src/dvc_data/fs.py

```diff
@@ -64,14 +64,22 @@ def _get_fs_path(self, path: "AnyFSPath"):
             if data:
                 fs, fs_path = data
                 if fs.exists(fs_path):
-                    return fs, fs_path
+                    return fs, typ, fs_path
 
         raise FileNotFoundError
 
     def open(  # type: ignore
         self, path: str, mode="r", encoding=None, **kwargs
     ):  # pylint: disable=arguments-renamed, arguments-differ
-        fs, fspath = self._get_fs_path(path, **kwargs)
+        cache_odb = kwargs.pop("cache_odb", None)
+        fs, typ, fspath = self._get_fs_path(path, **kwargs)
+
+        if cache_odb and typ == "remote":
+            from dvc_data.hashfile.build import _upload_file
+
+            _, obj = _upload_file(fspath, fs, cache_odb, cache_odb)
```

Comment:
This may be a bit inefficient API, we need something like `add()` that can stage as well as `add` to the odb in one go.

Reply:
@skshetry This does look convoluted. Why can't plots just `cache.add` a file they are already reading?

Reply:
Because it may be git-tracked or already cached.

Reply:
This `cache_odb` only works for `open`, which is odd (i would expect it to work for downloading too).

Reply:
Overall `cache_odb` looks very out of place in datafs 🙁

Reply:
yeah, I decided not to add to `get_file` at the moment.

Reply:
I personally find it out of place in `open()`, and it probably should be at the instance-level, as we always have a cache_odb (even if it's a readonly).

The `datafs` does not provide enough visibility on where the data is coming from.
This means it has to be implemented somewhere.